cpp/include/raft/neighbors/detail/ann_mg.cuh

```diff
@@ -0,0 +1,712 @@
+        root_handle, n_rows_per_batches, n_neighbors);
+
+      IdxT offset        = 0;
+      IdxT query_offset  = 0;
+      IdxT output_offset = 0;
+      for (IdxT batch_idx = 0; batch_idx < n_batches; batch_idx++) {
+        n_rows_per_batches   = std::min(n_rows_per_batches, n_rows - offset);
+        auto query_partition = raft::make_host_matrix_view<const T, IdxT, row_major>(
+          query_dataset.data_handle() + query_offset, n_rows_per_batches, n_cols);
+
+        RAFT_NCCL_TRY(ncclGroupStart());
+        for (int rank = 0; rank < num_ranks_; rank++) {
+          RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+          auto& ann_if = ann_interfaces_[rank];
+          ann_if.search(
+            dev_resources_[rank], search_params, query_partition, n_neighbors, root_rank_);
+
+          const auto& root_handle = set_current_device_to_root_rank();
+          const auto& comms       = resource::get_comms(root_handle);
+          uint64_t batch_offset   = rank * n_rows_per_batches * n_neighbors;
+          comms.device_recv(in_neighbors.data_handle() + batch_offset,
+                            n_rows_per_batches * n_neighbors,
+                            rank,
+                            resource::get_cuda_stream(root_handle));
+          comms.device_recv(in_distances.data_handle() + batch_offset,
+                            n_rows_per_batches * n_neighbors,
+                            rank,
+                            resource::get_cuda_stream(root_handle));
+        }
+        RAFT_NCCL_TRY(ncclGroupEnd());
+
+        auto in_neighbors_view  = raft::make_device_matrix_view<const IdxT, IdxT, row_major>(
+          in_neighbors.data_handle(), num_ranks_ * n_rows_per_batches, n_neighbors);
+        auto in_distances_view = raft::make_device_matrix_view<const float, IdxT, row_major>(
+          in_distances.data_handle(), num_ranks_ * n_rows_per_batches, n_neighbors);
+        auto out_neighbors_view = raft::make_device_matrix_view<IdxT, IdxT, row_major>(
+          out_neighbors.data_handle(), n_rows_per_batches, n_neighbors);
+        auto out_distances_view = raft::make_device_matrix_view<float, IdxT, row_major>(
+          out_distances.data_handle(), n_rows_per_batches, n_neighbors);
+
+        const auto& root_handle_ = set_current_device_to_root_rank();
+        auto h_trans = std::vector<IdxT>(num_ranks_);
+        IdxT translation_offset = 0;
+        for (int rank = 0; rank < num_ranks_; rank++) {
+          h_trans[rank] = translation_offset;
+          translation_offset += ann_interfaces_[rank].size();
+        }
+        auto d_trans = raft::make_device_vector<IdxT, IdxT>(root_handle_, num_ranks_);
+        raft::copy(d_trans.data_handle(), h_trans.data(), num_ranks_, resource::get_cuda_stream(root_handle_));
+        auto translations = std::make_optional<raft::device_vector_view<IdxT, IdxT>>(d_trans.view());
+        raft::neighbors::brute_force::knn_merge_parts<float, IdxT>(root_handle_,
```

Comment:
Out of scope for the current PR, but we might consider as a follow up: IVF-PQ or CAGRA-Q only return approximate distances. While merging parts based on the approximate distances, we might be throwing out good neighbors due to innacurate distance values. If we plan to do refinement, then we can treat the `in_neighbors` as candidates for refinement, and run refinement directly instead of calling `knn_merge_parts`.